src/librustc_typeck/check/generator_interior.rs

```diff
@@ -26,14 +26,27 @@ struct InteriorVisitor<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
     fcx: &'a FnCtxt<'a, 'gcx, 'tcx>,
     types: FxHashMap<Ty<'tcx>, usize>,
     region_scope_tree: Rc<region::ScopeTree>,
+    expr_count: usize,
 }
 
 impl<'a, 'gcx, 'tcx> InteriorVisitor<'a, 'gcx, 'tcx> {
     fn record(&mut self, ty: Ty<'tcx>, scope: Option<region::Scope>, expr: Option<&'tcx Expr>) {
         use syntax_pos::DUMMY_SP;
 
+        if self.fcx.tcx.sess.verbose() {
+        let span = scope.map_or(DUMMY_SP, |s| s.span(self.fcx.tcx, &self.region_scope_tree));
+        self.fcx.tcx.sess.span_warn(span, &format!("temporary scope for node id {:?}", expr));
+        }
+
         let live_across_yield = scope.map_or(Some(DUMMY_SP), |s| {
-            self.region_scope_tree.yield_in_scope(s)
+            self.region_scope_tree.yield_in_scope(s).and_then(|(span, expr_count)| {
+                // Check if the span in the region comes after the expression
+                if expr_count > self.expr_count {
```

Comment:
This makes me pretty uncomfortable. It's unclear that a "count" of the number of expressions is a good way of judging control-flow. I'd feel better if we used just the **statements** of a block: this would make simple stuff like

```
fn bar(baz: String) -> impl Generator<Yield=(), Return=()> {
    move || {
        yield drop(&baz);
    }
}
```

work fine, but wouldn't get the 'intra-statement' ordering. This seems .. kind of ok?  Or do you think it's not good enough?

Reply:
I suspect we'd want to be more flexible here. `await!` will probably be nested inside expressions, unlike plain `yield` which doesn't have a useful result.